### DISCUSSION TITLE
Grouping variants together

### DISCUSSION THREAD

**Discussion by siamekanto19** [tailwindlabs/tailwindcss#8337] [Ideas]
It would be great if we could group the classes used inside classes based on their variants. Now if we want to change two styles on hover state, we have to write `hover:bg-red-500 hover:text-white` and so on. It would be great if we could group these styles together for hover state so we could write something like `hover:(bg-red-500 text-white)`. We could group all other states like that too. It would really make the developer experience a lot better. Some other frameworks like WindiCSS, UnoCSS are already doing it. It would be great to see this in Tailwind too.

**Comment by randallmlough** [comment]
Huge +1 for this. Grouping would make reading tailwind classes so much easier. IMO readability goes down exponentially the more classes and variants of those classes gets mixed together. 

[Twin Marco](https://github.com/ben-rogerson/twin.macro) does a great job of grouping and even sub grouping. I would love to see this incorporated into tailwindcss natively 

EDIT: after reading the [Adam's twitter thread](https://twitter.com/adamwathan/status/1461519820411789314) I see they've researched this issue extensively. Hoping one day it becomes an addition without performance issues

**Comment by tsathis** [comment]
References: [Variant Groups in WindiCSS](https://windicss.org/features/variant-groups.html#variant-groups)

**Comment by milamer** [comment]
I wrote a transformer that parses the files before tailwind uses them to spread the grouped variants into single class names. This should remove the increased css bundle size.  
The drawback is that you have to wrap your class names into a spread function that will spread the variants increasing the runtime.

[tailwind-group-variant](https://github.com/milamer/tailwind-group-variant)

```ts
const expandVariant= createTransformer({ separatorChar: ' ' })
expandVariant('lg:(max-w-2xl w-full pb-2) xl:pb-32')
// lg:max-w-2xl lg:w-full lg:pb-2 xl:pb-32
```

```js
// tailwind.config.js
const createTransformer = require('tailwind-group-variant')
module.exports = {
  content: {
    files: ['./app/**/*.{ts,tsx,jsx,js,mdx}'],
    transform: createTransformer({ separatorChar: ' ' }),
  },
  ...
}
```

**Reply by ackvf** [reply]
I wish it was possible to use more seamlessly, without needing to import it everywhere. Does tailwind not support something like preprocessors?

**Reply by MrOxMasTer** [reply]
> I wish it was possible to use more seamlessly, without needing to import it everywhere. Does tailwind not support something like preprocessors?

https://github.com/MrOxMasTer/tailwindcss-multiple-classes

**Comment by Nefcanto** [comment]
The argument Adam made on Twitter is one of the worst arguments a developer could make.

I fail to understand why they did not consider separating grouped classes first and then compiling them.

This simple idea came to the mind of @milamer and I also thought about it. This should be very very easy, right?

**Reply by RobinMalfait** [reply]
Please do not assume that other people have "the worst argument", if you don't know the full story and don't know all the trade-offs.

---

> The argument Adam made on Twitter is one of the worst arguments a developer could make.

This can be explained by your next sentence:

> I fail to understand why they did not consider separating grouped classes first and then compiling them.

It's not that Adam is making "the worst argument", it's that you don't understand why there are limitations to this problem. And that is totally fine, I just want to give you some context about the trade-offs and issues in general so that you can change your "very very easy" statement to something like "oh I see".

Tailwind CSS is currently a build-time compiler that at the end of the day generates a CSS file based on your content files. If you use a class like this: `md:(font-bold,underline)` behind the scenes we would generate `md:font-bold md:underline` and then replace the selector again with your original selector. 

Why is that you may ask?

Well, if we just compile to:
```css
.md\:font-bold {}
.md\:underline {}
```

Then not a single selector matches the selector in your template. This means that no styles will be applied. To fix this there are a few solutions:

1. The solution we currently have, by replacing the individual rules with the selector you have in your template files.
2. Mutate source/generated files.
3. Introduce a (JS?) runtime.

I know you are hinting at the solution from @milamer which is a good solution and it works by:

1. introducing a JS runtime (`expandVariant` function)

This is fine for JS based projects and as long as the `expandVariant` isn't too expensive there should not be any (performance) problems (but please still measure if it is the case!). But we don't want to ship runtime code when it can be solved without runtime code. It isn't as expensive as some CSS-in-JS tools, but they have different trade-offs. However, there are products like [Maizzle](https://maizzle.com) that use Tailwind for email templates where it is impossible to run JavaScript and therefore also impossible to ship a runtime.

You are also hinting at "I fail to understand why they did not consider separating grouped classes first and then compiling them." we could do that, but then you have to rewrite source files (otherwise your selectors won't match), so all of a sudden you have to deeply integrate with your bundler (if you even use one).

This comes with its own problems, because Tailwind doesn't only work with HTML/JS based projects, it works with a bunch of [programming languages and frameworks](https://tailwindcss.com/docs/installation/framework-guides). You can use it in PHP, Ruby, Elixir, Rust, ... 

If we introduce something that can replace source code, then we have to have knowledge about all these programming languages to ensure it is safe to replace some code.

Even in HTML/JS, there are some problems:
```html
<div class="md:(font-bold,underline)"></div>
```
This might look easy, we can just find / replace.

How about comments?
```html
<!-- md:(font-bold,underline) -->
<div class="underline"></div>
```

Do we need to replace those? No? Well, now we need to ship an HTML parser so that we can _know_ about what parts are safe to replace and which parts should not be touched.

How about JS variables:
```js
const myVariable = 'md:(font-bold,underline)'
<div className={myVariables}></div>
```

How do we know that it is safe to mutate that `myVariable` variable? We would need yet another parser and know where the variable is used, know that it is used in the className, know about its scope to know that it is safe to mutate those.

But what if you call some function:

```js
const myVariable = myFunction('md:(font-bold,underline)')
<div className={myVariables}></div>
```

We don't really know what `myFunction` does. Is it safe to change `md:(font-bold,underline)` to just `md:font-bold md:underline`? We don't really know. Maybe it is statically analysable maybe it is not, it will highly depend on the project and on what `myFunction` does.

What else can we do? Well, we can think about some heuristics and only make it work for the cases we are 100% sure about that it would be safe. This is fine, but then you will run into situations where it works in some scenarios and not in others and you will have a very hard time trying to debug what's going on here.

Another common solution is to introduce something like:
```js
let classes = tw`md:(font-bold,underline)`
```
Where the `tw` can act as some marker where we know it is safe to use. However, this would only work in JS based projects not in projects with different programming languages like PHP, Elixir, Ruby, Rust, ...

---

So at the end of the day, if you want to introduce a runtime in your project by creating an `expandVariant` function or similar like @milamer described, by all means do it if it works for you. But I hope you understand why we can't "bake it in" because it is "very very easy" because it is simply not "very very easy".

**Reply by Nefcanto** [reply]
@RobinMalfait thank you so much for taking the time to explain in detail and my apologies for being harsh.

However, I still think that there is a very very easy way to do this.

Please let me explain.

As you mentioned, Tailwind is a build-time tool. Let's review why we use build tools in the first place. 

## Why do we use build tools? To build codes we write.

Just the way TypeScript is complied/transpiled to JS, Tailwind can replace `md:(text-red-500 font-bold)` with `md:text-red-500 md:font-bold` **in the output**, not in the source code.

This means that my source code is not changed at all, but inside the browser's dev tools, I will see the expanded version of my classes.

* This does not need any knowledge of the source programming language.
* It also does not require shipping any runtime stuff. 
* And it won't mess with the source code of developers. 
* And it's also intuitive.
* And it's very easy to implement (grouped classes detection and expansion is simple)

Please let me know if I could convey my solution, and please tell me what can be the problems with this approach.

**Reply by wongjn** [reply]
> - It also does not require shipping any runtime stuff.
> - And it won't mess with the source code of developers.

Unfortunately, one of these would have to be false. Let's take PHP for example, say I have some PHP code:
```php
echo '<div class="md:(mb-20,p-20)"></div>';
```
or
```php
function add_my_classes($classes) {
  $classes[] = 'md:(mb-20,p-20)';
  return $classes;
}
```
With the way many PHP applications are structured, this is the source code **and** the production code that would be on the server. Tailwind would have to transform this and thus *would* alter the source code. Or conversely, if it did not, a runtime would need to be shipped to transform the `md:(mb-20,p-20)` on the client.

**Reply by Nefcanto** [reply]
@wongjn Although some developers go on the production server and change files directly there, at the end of the day they should come back and apply those changes in the development code.

The PHP script on the production machine, **is the production code**. They either should not change it or in case it's urgent, the classes are still pretty straightforward.

I completely disagree with the point you mentioned.

**Reply by adamwathan** [reply]
@Nefcanto What you are suggesting is that Tailwind should become a compiler for your actual template files.

So say you are working on a PHP project, now you need to have all of your regular source code in one folder, and Tailwind needs to transform that source code and "build" it to another folder like `dist` or something, where we take a file like `welcome.php` and generate a new copy of `welcome.php` somewhere else that has the class names transformed.

To me this is just totally outside the scope of Tailwind and introduces a lot of complexity. So many programming environments are designed to execute the actual source files, not execute a separate compiled version of the source files generated by a CSS framework.

I don't know how this would even be possible in a Laravel project for instance — Tailwind would basically be responsible for cloning the entire project into a new gitignored folder and you'd have to run your dev server from the generated folder instead of the source files, which just seems extremely heavy handed and weird to me.

If part of using Tailwind meant making Tailwind the tool responsible for duplicating all of your code into a new compiled folder with the classes changed, my gut is that it would just be far too much for people to stomach just to generate some CSS, and the complexity of the project itself and surface area for potential bugs would increase dramatically.

Put another way, making Tailwind force users to go from this project folder structure:

```
app/
bootstrap/
config/
database/
lang/
public/
resources/
routes/
storage/
tests/
├─ composer.json
├─ package.json
```

...to this:

```
├─ dist (copy of code generated by Tailwind with classes expanded)
   ├─ app/
   ├─ bootstrap/
   ├─ config/
   ├─ database/
   ├─ lang/
   ├─ public/
   ├─ resources/
   ├─ routes/
   ├─ storage/
   ├─ tests/
├─ src
   ├─ app/
   ├─ bootstrap/
   ├─ config/
   ├─ database/
   ├─ lang/
   ├─ public/
   ├─ resources/
   ├─ routes/
   ├─ storage/
   ├─ tests/
├─ composer.json
├─ package.json
```

...just to expand some class names feels like an extremely bad trade-off in complexity to me.

**Reply by gustavopch** [reply]
> But we don't want to ship runtime code when it can be solved without runtime code.

Late to the party, but well... it seems it's just not possible to solve without runtime code, so maybe just ship runtime code?

**Comment by Nefcanto** [comment]
@adamwathan that was a compelling argument. Yeah, you're right. Doing all that to accommodate class name expansion is a bad decision. But I didn't mean Tailwind to become a compiler. I meant somehow to hook into existing compilers and do some work. Some pre-compile work.

However, I still think we need to find a way to expand our classes somehow. Why do I say that? 

It's because we're using Tailwind for literally every project we have. It's now about a year I guess that we have not written one line of CSS in any other form. That's how good your invention is. And I'm truly thankful for it. It helps a lot with clean code + rapid development.

Yet that `md:(text-red-500,font-bod)` thing is very ugly and ruins the spirit of Tailwind which would make our codes clean.

My recommendation is that we have a simple **run-time** that is not enabled by default and we could configure it to opt-in.

For example in my `tailwind.config.js`:

```
module.exports = {
    includeClassExpansionRuntime: true
}
```

Then in my layout code:

```
import { classExpansionRuntime } from 'tailwindcss`
// then somehow call it:
// classExpansionRuntime.expandClasses();
```

Then in our code, we can write:
```
md:(text-red-500 font-bold)
```

**Comment by stephenkoo** [comment]
I'm really hoping this gets implemented due to long chained states. 

E.g.

```
supports-[not(container-type:inline-size)]:min-height-[10px]:hover:bg-red-200 supports-[not(container-type:inline-size)]:min-height-[10px]:hover:border-red-500 supports-[not(container-type:inline-size)]:min-height-[10px]:hover:text-red-500
```

Could simply be written as:
```
supports-[not(container-type:inline-size)]:min-height-[10px]:hover:(bg-red-200 border-red-500 text-red-500)
```

**Reply by yf-hk** [reply]
Sorry for a bit of off topic, may I ask what syntax is `min-height-[10px]`?

**Reply by robbyrice** [reply]
That is the arbitrary value syntax https://tailwindcss.com/docs/adding-custom-styles#using-arbitrary-values

**Comment by stephenkoo** [comment]
I've been thinking, would this be feasible & desirable syntax for longer more complex tailwind classes? (Inspired by styled-components)

```jsx

import { tw } from 'tailwind'

const Component = () => {
  return (
    <div className={tw`
      bg-red-500 text-white p-4 rounded

      supports-[not(container-type:inline-size)]:
        w-1/2 mx-auto sm:w-full sm:mx-0

        min-height-[10px]:
          h-10 sm:h-20 md:h-30 lg:h-40 xl:h-50

          hover:
          focus:
            bg-blue-500 text-black p-8 rounded
    `}>
       Hi
    </div>
  )
}
```

If done with tailwind currently, it would look like:

```jsx
const Component = () => {
  return (
    <div className="
      bg-red-500
      text-white
      p-4
      rounded

      supports-[not(container-type:inline-size)]:w-1/2
      supports-[not(container-type:inline-size)]:mx-auto
      supports-[not(container-type:inline-size)]:sm:w-full
      supports-[not(container-type:inline-size)]:sm:mx-0

      supports-[not(container-type:inline-size)]:min-height-[10px]:h-10
      supports-[not(container-type:inline-size)]:min-height-[10px]:sm:h-20
      supports-[not(container-type:inline-size)]:min-height-[10px]:md:h-30
      supports-[not(container-type:inline-size)]:min-height-[10px]:lg:h-40
      supports-[not(container-type:inline-size)]:min-height-[10px]:xl:h-50

      supports-[not(container-type:inline-size)]:min-height-[10px]:hover:bg-blue-500
      supports-[not(container-type:inline-size)]:min-height-[10px]:hover:text-black
      supports-[not(container-type:inline-size)]:min-height-[10px]:hover:p-8
      supports-[not(container-type:inline-size)]:min-height-[10px]:hover:rounded"

      supports-[not(container-type:inline-size)]:min-height-[10px]:focus:bg-blue-500
      supports-[not(container-type:inline-size)]:min-height-[10px]:focus:text-black
      supports-[not(container-type:inline-size)]:min-height-[10px]:focus:p-8
      supports-[not(container-type:inline-size)]:min-height-[10px]:focus:rounded"
    >
      Hi
    </div>
  )
}
```

To address [Adam's findings](https://twitter.com/adamwathan/status/1461519820411789314) of group variants causing more html/css bloat, the `styled` tagged template literal simply compiles to regular tailwind classes like above.

I've read @RobinMalfait 's [comment](https://github.com/tailwindlabs/tailwindcss/discussions/8337?sort=new#discussioncomment-4036320) with the example
```js
let classes = tw`md:(font-bold,underline)`
```
and am aware this probably would benefit only JS projects, but I think it would greatly benefit those who do use JS.

**Reply by coycoylaniba** [reply]
+1

**Comment by andershagbard** [comment]
Semi related to: https://github.com/tailwindlabs/tailwindcss/discussions/10925

**Comment by stephenkoo** [comment]
Just curious, how have others dealt with multi-chained states and queries in Tailwind?
I find it hard to read and debug tailwind classes once dealing with non-trivial scenarios that combines various factors like:

- varying styles based on variant & intent props (aware cva may be a solution here)
- multiple queries & states (e.g. using `supports`, `prefers-color-scheme`, `@container` & `:disabled/:hover` together for multiple styles)
- fallback styles e.g. (a) applying `height-[100svh]` with a fallback of `height-[100vh]` and (b) applying the same 30+ styles to when `@container sidebar (min-width: 400px)` as well as fallback `@media (min-width: 900px)` when `supports-[not(container-type:size)]` ?

Aware these may be contrived examples but they're not far off from my experience using css (e.g. i want to style my buttons to handle a combination of motion preference, dark/light modes, hover/active/disabled states, variant (solid, outline) + intent (error, success, primary), width at particular container queries with a media query fallback.)

At some point, grouping variants together may not be a sufficient solution to aid readability without the clarity of an indented/nested CSS structure to easily "drill-down" to what styles apply to a given state of a component.

**Comment by Toysoldier34** [comment]
This would be a great feature, even if it is something we have to opt into with a note about how not all systems may play well with it.

Tailwind quickly becomes a pain to read and organize once there are many conditional class styles being applied. I have not found a great solution for this other than moving to something that isn't Tailwind, or trying to implement one of these custom workarounds to add the feature ourselves. I have seen regular enough requests for the feature over the years, it is a shame there isn't an official option to help reduce redundant code that can really bloat the class names in our files as it would be a great quality of life improvement.

**Comment by Nefcanto** [comment]
@stephenkoo your solution is great. The point is that the Tailwind team has gone proud (no offense, just saying what I see). They do not listen to the community anymore. It happens to a lot of teams over the years. They started something extremely amazing, a revolution. But that made them think that they know **everything** and the rest of the community knows nothing. That's why we're here now.

We use Tailwind for 100% of our projects. We do not have one line of CSS or LESS or SASS or CSS-in-JS or styled-components or any other tech.

Yet we have experienced extremely dirty, unmaintainable code as you mentioned. When we want to do combine things, the current solution is simply not enough.

**Comment by sarimarton** [comment]
@adamwathan What is the downside of supporting the grouping syntax on Tailwind side, and leaving the template transformation  responsibility to userland? Tailwind could simply let the door open for the ecosystem to accommodate to it. This would be an additive feature, not impacting the current status quo. To make this feature work, we need involvement from both the ecosystem and Tailwind-side. Tailwind could simply (even silently, or not loudly communicated) do its part.

I understand that it might be confusing that Tailwind provides something which doesn't do anything on its own, but it would be more like an "enabler" feature. But seeing the demand and that how intuitive this would be anyway, I think it'd be worth. I also see the possible pitfall of having double parsers (one from Tw, one from ecosystem) which does the distribution. But given the relative simplicity of the syntax, at least on a conceptual level, I still think it could be worth.

**Comment by silicakes** [comment]
Thanks to @wongjn , this can now be done relatively easy:
https://github.com/tailwindlabs/tailwindcss/discussions/11701#discussioncomment-6569866

Kudos @wongjn !

**Comment by MrOxMasTer** [comment]
It's sad that we can't be given this feature in any form. Because in reality it is not convenient to write it and debug it too.

**Reply by gustavopch** [reply]
There's this plugin: https://github.com/brandonmcconnell/tailwindcss-multi

**Reply by katerlouis** [reply]
While obviously better than nothing, to me this still is a workaround. Correct me if I'm wrong, but when I use this plugin, I lose intellisense while writing inside the multi directive. Auto complete suggestions to see whats at offer and help use tailwind immensely, especially at the start of learning it.

**Comment by alanxp** [comment]
Would love to see this feature to come up, would save lots of headaches upon debuging components or even editing them. Not so much text would be nicer

**Comment by mordv** [comment]
Man, its so sad. The absence of this feature makes the exact mess as the lib intends to solve.
I've tldr'ed all the arguments against this feature, as the mentioned above kind of mess simply boldly sucks.

``` js
      supports-[not(container-type:inline-size)]:min-height-[10px]:h-10
      supports-[not(container-type:inline-size)]:min-height-[10px]:sm:h-20
      supports-[not(container-type:inline-size)]:min-height-[10px]:md:h-30
      supports-[not(container-type:inline-size)]:min-height-[10px]:lg:h-40
      supports-[not(container-type:inline-size)]:min-height-[10px]:xl:h-50
```

**Reply by gustavopch** [reply]
There's hope: https://x.com/adamwathan/status/1849509712368226792

**Reply by MrOxMasTer** [reply]
> There's hope: https://x.com/adamwathan/status/1849509712368226792

Feature request from 2018...

**Reply by MrOxMasTer** [reply]
You can try the new syntax today))) https://www.npmjs.com/package/tailwindcss-multiple-classes

**Reply by MrOxMasTer** [reply]
> Man, its so sad. The absence of this feature makes the exact mess as the lib intends to solve. I've tldr'ed all the arguments against this feature, as the mentioned above kind of mess simply boldly sucks.
> 
> ```js
>       supports-[not(container-type:inline-size)]:min-height-[10px]:h-10,sm:h-20,md:h-30,lg:h-40,xl:h-50
> ```

It doesn't really work with your example, but if you want, you can adjust it and it will work. I could fix it, but for now I'm too lazy ahah

**Reply by MrOxMasTer** [reply]
> Man, its so sad. The absence of this feature makes the exact mess as the lib intends to solve. I've tldr'ed all the arguments against this feature, as the mentioned above kind of mess simply boldly sucks.
> 
> ```js
>       supports-[not(container-type:inline-size)]:min-height-[10px]:h-10
>       supports-[not(container-type:inline-size)]:min-height-[10px]:sm:h-20
>       supports-[not(container-type:inline-size)]:min-height-[10px]:md:h-30
>       supports-[not(container-type:inline-size)]:min-height-[10px]:lg:h-40
>       supports-[not(container-type:inline-size)]:min-height-[10px]:xl:h-50
> ```

https://www.npmjs.com/package/tailwindcss-multiple-classes

now my plugin supports such classes). Of course, I haven't experienced it in all situations, but I think it's already good

**Reply by MrOxMasTer** [reply]
> Man, its so sad. The absence of this feature makes the exact mess as the lib intends to solve. I've tldr'ed all the arguments against this feature, as the mentioned above kind of mess simply boldly sucks.


I have updated the plugin. It is now available for Webpack/next.js/vite/rollup/PostCSS

**Comment by xsjcTony** [comment]
Based on the content above, I'm wondering why can't we implement it for specific build tools?
E.g. `UnoCSS`'s  `transformVariantGroup` transformer works well for `Vite`, but won't work with `PostCSS` + `Next.js` setup.
This is totally acceptable, as those guys using Vite can benefit from this feature, while those using Next.js can't.

Same for Tailwind here, build this feature for those Vite users (which I believe it's now the majority of web projects), and just mention this won't work for Next.js, PHP, Ruby, etc.

I don't really see the issue there, the simple theory is, if you can benefit from it, then use it, otherwise go for the traditional way.

Some other framework have successfully done this, at least for the JS ecosystem, so I don't see why Tailwind can't do so.

**Comment by FarhanXTanvir** [comment]
After skimming through various arguments of this thread what I understood is tailwind team has not yet found an optimal solution for adding this feature as they are thinking of providing multi language support. But why don't they think of language specific supports too? If PHP or Ruby can't afford to have this feature then it's the drawback of that language but for that reason why should js based lib/framework be deprived of having this wonderful feature? you can simply add a config to enable this features in js. You can make this feature strongly opinionated for your convenience. You can also provide the support for user preferred regex of grouping variants.

**Comment by alanxp** [comment]
as of tailwind 4.0 we still don't see this feature coming....

**Comment by gustavopch** [comment]
If I understand correctly, the main problem that this would introduce is: each class group would be a new class in the generated CSS, so it would quickly bloat the generated CSS with classes that are used only once, harming one of the main benefits of TailwindCSS.

BUT... TailwindCSS v4 has a Vite plugin, so maybe it doesn't need to actually generate classes for the groups?

I mean, if the source code has `class="hover:(bg-red-500;text-white)"`, instead of adding `.hover\:\(bg-red-500\;text-white\)` to the generated CSS, it could simply transform the source code to `class="hover:bg-red-500 hover:text-white"` and add those 2 separate classes.

This would only work with the Vite plugin as it would need to transform the source code, but that's fine, each way to use TailwindCSS has its pros and cons.

**Comment by kfdf** [comment]
It's 2025 and all major browsers have been supporting css nesting for more than a year now. Here is a solution (proof-of-concept more like) that makes use of it.

https://github.com/kfdf/tailwind-modifier-aliasing

**Comment by markchitty** [comment]
Possibly blasphemy on a Tailwind discussion, but [UnoCSS](https://unocss.dev/guide/why) supports [variant groups](https://unocss.dev/transformers/variant-group) out of the box, along with a bunch of other [cool utilities](https://unocss.dev/presets/attributify) including [Tailwind compatibility](https://unocss.dev/presets/wind#rules). Haven't tried it yet but it could be a viable alternative?

**Comment by jimydavis** [comment]
I wrote this plugin that helps me a lot for the nature of grouping responsive modifiers: https://www.npmjs.com/package/vite-plugin-twneat 

It probably has bugs and not very performant, but the code is fairly straightforward - it just concats things and dumps it in a safelist. I am not sure but I don't think it produces more classes than what you would have written natively without groupings.

PS: I am a backend developer who started writing some simple landing page code and got overwhelmed with the entire javascript ecosystem. So apologies if I don't understand what is going on.

**Reply by MrOxMasTer** [reply]
> I wrote this plugin that helps me a lot for the nature of grouping responsive modifiers: https://www.npmjs.com/package/vite-plugin-twneat
> 
> It probably has bugs and not very performant, but the code is fairly straightforward - it just concats things and dumps it in a safelist. I am not sure but I don't think it produces more classes than what you would have written natively without groupings.
> 
> PS: I am a backend developer who started writing some simple landing page code and got overwhelmed with the entire javascript ecosystem. So apologies if I don't understand what is going on.

https://github.com/MrOxMasTer/tailwindcss-multiple-classes

**Reply by jimydavis** [reply]
Read your very detailed explanation in the other thread! Thanks for the illumination.

**Comment by wffranco** [comment]
Maybe with braces + commas, we can group items and/or variants.
```html
<button class="bg-blue-500 text-white {hover,focus}:{bg-blue-300,text-gray-200} dark:{bg-gray-100,text-gray-600,{{hover,focus}:{bg-gray-300,text-gray-700}}">
  Using brackets and commas
</button>
```
- Currently you're using `()` and `[]` in different cases, then using `{}` it will not crash or affect the current behavior.
- We can group variants, like `{hover,focus}:opacity-100` that should apply both variants but separated (one for hover and one for focus, not nested).
- We can subgroup variants, like `dark:{hover,focus}:hidden`.
- We can group items, like `sm:{m-0,p-0,block}`.

**Reply by gustavopch** [reply]
I like how this syntax resembles Bash's [brace expansion](https://www.gnu.org/software/bash/manual/html_node/Brace-Expansion.html).

**Reply by wffranco** [reply]
I know one of the main discussions is about repeated css, but if you think about it, when you have a very complex behaviors, with a lot of variants, we have currently 2 options: adding one by one each style, or creating components.
And if you think about it, create a component for a single use is not better than grouping.
Also in tailwind 4 we have a problem in frameworks when we need to use `@apply` outside the main tailwind css file.

**Comment by ArturBaybulatov** [comment]
FYI, there's a library called [typewind](https://github.com/Mokshit06/typewind) for type-safe Tailwind, which also allows grouping:

```ts
import { tw } from 'typewind';

const mediaStyles = tw.sm(tw.w_4.mt_3).lg(tw.w_8.mt_6);
const groupInGroups = tw.text_sm.sm(tw.bg_black.hover(tw.bg_white.w_10));
```

**Reply by ackvf** [reply]
Pretty neat. I built a primitive [coloring](https://gist.github.com/ackvf/68f992660a5eda645c4671d3599b2acf#file-colorizer-proxy-ts) "library" with Proxy that you could chain indefinitely some years ago. This project gives me nostalgia.

**Comment by FireyFly89** [comment]
This topic is still ongoing? I have alternative suggestions, I mean the main focus here is responsivity as far as I'm aware, right? So what if we could do lg:my-custom-class and have all the rules under my custom class that I've made only be applied above the set resolution. I don't think I can do this at the moment, or if I can I'm sorry I was not aware, but I've tested it, and it does not work. If I could do something like let's say I have 2 classes called "block, block-compact" and I'd do something like "max-lg:block-compact:lg:block" perhaps not the best syntax idea, I just wanted to throw the idea here, because this way those who like to make custom classes in components would have a powerful tool to handle responsivity. My specific problem all the time with the mobile first development is that I have to override the mobile rules all the time, but what if we're just using both max and min breakpoints correctly, and not overusing max breakpoints? To me it makes sense.

**Comment by a1um1** [comment]
I think tailwindcss is on life support now

**Comment by rozsazoltan** [comment]
> There's an interesting analysis about this - I'll pull it up for you in a moment. The conclusion is that no, and it wouldn't be very useful anyway.
> 
> * https://github.com/tailwindlabs/tailwindcss/discussions/8337#discussioncomment-4036320
> * https://x.com/adamwathan/status/1461519820411789314
> * https://x.com/adamwathan/status/1461519824828473359
> * https://x.com/adamwathan/status/1461519828708212737
> * https://x.com/adamwathan/status/1461699008682934275
>
> Source: https://discord.com/channels/486935104384532500/1416452801487441930/1416453985635799153

The only advantage is avoiding the continuous typing of variants. But it has many drawbacks. For one, TailwindCSS pre-generates the styles, so a class like `.md(text-2xl,red,p-3) { ... }` would need to exist. Many such combinations would likely only occur once, resulting in a lot of unnecessary class declarations, which greatly increases the size of the generated CSS and leads to higher network traffic.

In contrast, `.md:text-2xl { ... } .md:red { ... } .md:p-3 { ... }` is more likely to be reused multiple times. Even if not, you won't end up with many separate duplicates for different groupings like `.md(text-2xl,red,p-3) { ... }`, `.md(text-2xl,p-3,red) { ... }`, or `.md(text-2xl,red) { ... }`.
* https://x.com/adamwathan/status/1461519824828473359

In order for the generated CSS not to have to include `.md(text-2xl,red,p-3) { ... }` explicitly, the name `md(text-2xl,red,p-3)` would need to be resolved by some function into `md:text-2xl md:red md:p-3`. However, for TailwindCSS to currently interpret these, it would also need to resolve the shorthand. This, in turn, would require using a JS function unnecessarily to parse every class name list, which could add extra runtime to your code - essentially for no real benefit.

And at the end of the day, you'd only save a few characters with this mega-grouping:
```none
md:text-2xl md:red md:p-3
md(text-2xl,red,p-3)
```
* https://x.com/adamwathan/status/1461519820411789314

**Reply by sarimarton** [reply]
> you'd only save a few characters

With selector modifiers this 'few characters' would be a serious bloat cut.

**Reply by rozsazoltan** [reply]
@sarimarton yeah, there are cases where you can save more than few characters.
```md
md:text-2xl md:text-black md:bg-white md:p-3 md:dark:text-white md:dark:bg-black
md:(text-2xl,text-black,bg-white,p-3,dark(text-white,bg-black))
```

In exchange, you could go with two approaches:
* A. TailwindCSS interprets these groups itself and generates the CSS. In this case, you'd have to produce many more classes, even if the individual classes already exist, because the group would require merging as well.
  * I think that's completely unthinkable - the size of the generated CSS is very important, especially when traffic is high.
  * See: https://x.com/adamwathan/status/1461519824828473359 and https://x.com/adamwathan/status/1461699008682934275
* B. (1.) Don't generate CSS for the group itself, but have TailwindCSS resolve the group strings and create CSS for the expanded class names. (2.) And for runtime this would require a JS function to run for every class declaration on each page load, resolving all groups back to their original meaning so they match the compact generated CSS. This adds extra runtime overhead. (I'd personally only implement this if I really wanted to, but even then it adds extra load compared to simply typing the class names statically during development without using groups.)
  * Of course, it doesn't affect TailwindCSS generating the CSS during the build, but afterward, even the smallest unnecessary delay when loading my project matters.
  * See more: https://github.com/tailwindlabs/tailwindcss/discussions/8337#discussioncomment-4036320
* C. I would keep part (1.) of option B; and for the (2.) part, I'd resolve the groups once during the build, so that in production they appear without the groups. I'm not sure what extra challenges this might entail, but maybe it's feasible? However, this would only be feasible with a bundler, so it's out of the question for CLI-only projects. I'll sleep on this a few nights, but the idea isn't dismissible.

**Reply by rozsazoltan** [reply]
Btw, if you want to shorten selectors, you can still declare new variants using `@custom-variant` for any selector you use so frequently that it becomes annoying.

**Comment by 3xau1o** [comment]
UnoCSS also [has support for variant groups](https://unocss.dev/transformers/variant-group#variant-group-transformer) and many other stuff officially

**Comment by kerryj89** [comment]
Funny enough, AI hallucinated this feature for the first time today.  I got really excited for a second considering how much this can clean parts of the code.  But alas, it was not a feature.  I find hallucinations presenting features and cli options that you didn't ask for to be a good indicator that _maybe_ it should be considered.

<img width="763" height="658" alt="CleanShot 2026-04-16 at 16 22 04" src="https://github.com/user-attachments/assets/c6adaf12-4c95-4efa-97a1-ec1b3d94eba2" />

**Reply by rozsazoltan** [reply]
* https://github.com/tailwindlabs/tailwindcss/discussions/8337#discussioncomment-4036320
* https://github.com/tailwindlabs/tailwindcss/discussions/8337#discussioncomment-14393233
* https://github.com/tailwindlabs/tailwindcss/discussions/8337#discussioncomment-14395170
